### ISSUE TITLE
Session.id unit test failures 

### ISSUE THREAD

**Issue by mmd-osm** [openstreetmap/openstreetmap-website#2488]
Could the commit below (security fix for CVE-2019-16782) be the reason why our unit tests are failing? We’re using session.id in a header erb and it triggers an error when generating the logout link. I’m seeing this `request.session.id&.public_id` in one of the unit tests: 

https://github.com/rails/rails/commit/92ec9f270d696a17ba7604b3d02931f07c3e7c6a#diff-ddd3ea55daa7e105c86ed353964cfd5b

**Comment by tomhughes** [comment]
Yes I've been looking at it this evening and got as far as working out that `session.id` sometimes fails with the latest `rack` but I'm still working on the details...

**Comment by tomhughes** [comment]
So the story is that `session.id` still works when using the memcache session store but fails when using the in memory cache which is the default - in that case you need `session.id.public_id` but that fails in memcache mode!

So it's working for me locally as I have memcache configured. Travis is also supposed to have it configured but apparently it isn't working...

**Comment by tomhughes** [comment]
I've fixed the travis config in c524cbf096d9f67936a5ec6050e5c83fd49077cd so that it uses memcache as intended which has fixed travis but it is still broken for developers using the default settings.

It's hard to fix until upstream fixes that CVE for memcache, or at least makes `public_id` work for memcache.

**Comment by mmd-osm** [comment]
Thanks! So if anyone comes across the following Runtime error in  `<%= link_to t("layouts.logout"), logout_path(:session => session.id, :referer => request.fullpath), :class => "geolink" %>
`
```
RuntimeError - :
  app/views/layouts/_header.html.erb:105
  app/views/layouts/site.html.erb:5
  app/views/layouts/map.html.erb:87
  app/controllers/application_controller.rb:321:in `better_errors_allow_inline'
  config/initializers/compressed_requests.rb:27:in `call'
  config/initializers/cors.rb:9:in `call'

```

you need to enable memcached by adding the following to config/settings.local.yml and enabling the memcached service:

```
memcache_servers: 
                  - 127.0.0.1
```

**Comment by tomhughes** [comment]
Well you also need to have memcached running ;-)

**Comment by Legolash2o** [comment]
How do you setup/enable the memcached service?

**Comment by gravitystorm** [comment]
I've been digging into this, since I found it strange that we need to use the session id in logout links in the first place. I found that the reason was to avoid people being logged out by simple get requests: https://trac.openstreetmap.org/ticket/2792 (10 years ago).

But why is the logout link a GET request in the first place? It changes state on the server, and is effectively "DELETEing a user session" (this is how Devise models it, for example). So I think the real fix might be to change the link and the action on the logout method, and that will mean we can avoid any worrying about the internal session id too.

**Comment by tomhughes** [comment]
Well I suspect it is just a consequence of the fact that it's a link, not a button, and GET is the natural (indeed only bar rails trickery) method available to a link.

I would also add that there are other places in the code that use `session.id` and which will have the same problem.

**Comment by josetuna** [comment]
Hi to all! I'm trying to install Rails Port on an Ubuntu 18.04 VM. I've followed the tutorial, but fails in the test with this kind of errors:

```
Error:
UserDiariesTest#test_showing_create_diary_entry:
ActionView::Template::Error: 
    app/views/layouts/_header.html.erb:105
    app/views/layouts/site.html.erb:5
    app/controllers/diary_entries_controller.rb:21:in `new'
    config/initializers/compressed_requests.rb:27:in `call'
    config/initializers/cors.rb:9:in `call'
    test/integration/user_diaries_test.rb:29:in `test_showing_create_diary_entry'
```
All errors have almost the same call stack: every step which calls function `call `in `cors.rb` fails.
¿Could it be related to this issue?

I installed memcached and modified settings.local.yml as suggested but still no success...
Thanks in advance!

**Comment by gravitystorm** [comment]
@josetuna If you update to the latest commit, this will be fixed. It was fixed in PR #2496
